### PR TITLE
Core(M): Add support for LLVM/Clang

### PR DESCRIPTION
**Pull request by JonatanAntoni**
- Add compiler header for clang.
- Enhance CoreValidation for clang.

### CODE REVIEW

File: .github/workflows/corevalidation.yml
Comment (GuentherMartin) [inline .github/workflows/corevalidation.yml:1]: Can not judge,

File: CMSIS/CoreValidation/Project/vcpkg-configuration.json
Comment (GuentherMartin) [inline CMSIS/CoreValidation/Project/vcpkg-configuration.json:1]: Can not judge

**Review by GuentherMartin — DISMISSED** [review]
Looks fine to me

File: CMSIS/Core/Include/cmsis_clang.h [hunk 1]
```diff
@@ -0,0 +1,2073 @@
...
+__STATIC_FORCEINLINE uint32_t __RRX(uint32_t value)
+{
+  uint32_t result;
+
+  __ASM volatile ("rrx %0, %1" : __CMSIS_GCC_OUT_REG (result) : __CMSIS_GCC_USE_REG (value) );
+  return(result);
+}
+#endif /* __ARM_ARCH_ISA_THUMB >= 2 */
+
+#ifdef __ARM_FEATURE_LDREX
```
Comment (GuentherMartin) [inline CMSIS/Core/Include/cmsis_clang.h:465]: Change 
#ifdef __ARM_FEATURE_LDREX
to
#if __ARM_FEATURE_LDREX >= 7

see https://developer.arm.com/documentation/101028/0012/5--Feature-test-macros

File: CMSIS/Core/Include/cmsis_clang.h [hunk 2]
```diff
@@ -0,0 +1,2073 @@
...
+  \brief   Remove the exclusive lock
+  \details Removes the exclusive lock which is created by LDREX.
+ */
+__STATIC_FORCEINLINE void __CLREX(void)
+{
+  __ASM volatile ("clrex" ::: "memory");
+}
+
+/**
+  \brief   LDRT Unprivileged (8 bit)
```
Comment (GuentherMartin) [inline CMSIS/Core/Include/cmsis_clang.h:570]: are the LDR*/STR* Unprivileged instructions also covered by __ARM_FEATURE_LDREX?

**Review by GuentherMartin — CHANGES_REQUESTED** [review]

File: CMSIS/Core/Include/cmsis_clang.h [hunk 3]
```diff
@@ -0,0 +1,2073 @@
...
+#endif
+
+/* #########################  Startup and Lowlevel Init  ######################## */
+
+#ifndef __PROGRAM_START
+#define __PROGRAM_START           _start
+#endif
+
+#ifndef __INITIAL_SP
+#define __INITIAL_SP              __stack
```
[the remark this answers is not in the document]
Comment (JonatanAntoni) [reply CMSIS/Core/Include/cmsis_clang.h:127]: This is a linker defined symbol, defined in the linker script. We can use same names as in GCC if we adopt the linker script. I kept the symbols as defined in the toolchain provided linker script template.

[the remark this answers is not in the document]
Comment (GuentherMartin) [reply CMSIS/Core/Include/cmsis_clang.h:127]: if we provide the linker script than I would suggest to align the symbol names.

File: CMSIS/Core/Include/cmsis_clang.h (hunk 2 above)
Comment (JonatanAntoni) [reply CMSIS/Core/Include/cmsis_clang.h:570]: `LDRT<c>` is Thumb-2 instruction, so nothing to do with exclusive access ones.

File: CMSIS/Core/Include/cmsis_clang.h [hunk 4]
```diff
@@ -0,0 +1,2073 @@
...
+  \brief   Reverse bit order of value
+  \details Reverses the bit order of the given value.
+  \param [in]    value  Value to reverse
+  \return               Reversed value
+ */
+__STATIC_FORCEINLINE uint32_t __RBIT(uint32_t value)
+{
+  uint32_t result;
+
+#if __ARM_ISA_THUMB >= 2
```
[the remark this answers is not in the document]
Comment (JonatanAntoni) [reply CMSIS/Core/Include/cmsis_clang.h]: Done

File: CMSIS/Core/Include/cmsis_clang.h (hunk 1 above)
Comment (JonatanAntoni) [reply CMSIS/Core/Include/cmsis_clang.h]: Done

File: CMSIS/Core/Include/cmsis_clang.h (hunk 2 above)
Comment (JonatanAntoni) [reply CMSIS/Core/Include/cmsis_clang.h:570]: Done

**Pull request merged**